### DISCUSSION TITLE
Tips for managing FastAPI project

### DISCUSSION THREAD

**Discussion by yareyaredesuyo** [fastapi/fastapi#2491] [Show and tell]
It maybe useful to share the knowledge of fastapi's tips and practice.

## Set up FastAPI minimum project from scratch

This is a minimum setup procedure for a fastapi project using `poetry`.

- Use `poetry` for dependency management
  - ~`fastapi` itself is also using `poetry`~
    ~- https://github.com/tiangolo/fastapi/blob/master/pyproject.toml~
  - fastapi is using [flit](https://github.com/takluyver/flit) for the development dependencies:
    - https://fastapi.tiangolo.com/contributing/
- initialize project using poetry
  - `poetry new app && cd app`
- add fastapi related packages
  - `poetry add fastapi uvicorn`
- add `main.py` (touch app/main.py) and write min fastapi code, then start project.
  - `poetry shell`
  - `uvicorn app.main:app --reload`
- (Option) add lint, format, test related packages
  - `poetry add -D mypy flake8 autoflake isort black pytest-cov`
- (Option)add `scripts` folder and files.
  - implement `test`, `check`, `lint`, `coverage` files for convenience.
    - This is what `encode` libraries is doing things.
      - https://github.com/encode/starlette/blob/master/scripts/README.md
        - Styled after GitHub's ["Scripts to Rule Them All"](https://github.com/github/scripts-to-rule-them-all).
    - if you prefer fastapi's scripts, use it.
      - https://github.com/tiangolo/full-stack-fastapi-postgresql/tree/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/scripts
      - https://github.com/tiangolo/fastapi/tree/master/scripts
- (Option)add `python-dotenv` for management settings
- (Option)add `pre-commit` for formatting, linting, testing before commit
- (Option)add .gitignore from .gitignore/python folders
   - `wget https://raw.githubusercontent.com/github/gitignore/master/Python.gitignore -O .gitignore`

minimum `pyproject.toml's` example

```toml
[tool.poetry]
name = "app"
version = "0.1.0"
description = ""
authors = ["admin <admin@example.com>"]

[tool.poetry.dependencies]
python = "^3.9"
fastapi = "^0.62.0"
uvicorn = "^0.12.3"

[tool.poetry.dev-dependencies]
pytest = "^5.2"

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

Refs:
- https://dev.to/alexvanzyl/fastapi-simple-application-structure-from-scratch-2mem

## Folder structure

I still does know what is the best way for structuring fastapi project, but `full-stack-fastapi-postgresql` maybe helpful for thinking about folder structure.

```
└── app
    ├── alembic
    │   └── versions
    ├── app
    │   ├── api
    │   │   └── api_v1
    │   │       └── endpoints
    │   ├── core
    │   ├── crud
    │   ├── db
    │   ├── email-templates
    │   │   ├── build
    │   │   └── src
    │   ├── models
    │   ├── schemas
    │   └── tests
    │       ├── api
    │       │   └── api_v1
    │       ├── crud
    │       └── utils
    └── scripts
```

Refs:
- https://github.com/tiangolo/full-stack-fastapi-postgresql/tree/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app

## Useful resources

- fastapi docs - external-links
  - https://fastapi.tiangolo.com/external-links/
- github topics
  - https://github.com/topics/fastapi
- awesome fastapi
  - https://github.com/mjhea0/awesome-fastapi
- dev.to
  - https://dev.to/t/fastapi
- reddit
  - https://www.reddit.com/r/FastAPI/
- medium
  - https://medium.com/search?q=fastapi
- discord
  - https://discord.com/invite/VQjSZaeJmf
- stackoverflow
  - https://stackoverflow.com/questions/tagged/fastapi

**Comment by ycd** [comment]
Thank you for the post @yareyaredesuyo.

I just want to clarify a few things for future readers.

> fastapi itself is also using poetry
     Ref: https://github.com/tiangolo/fastapi/blob/master/pyproject.toml

Actually, this is not right,  `pyproject.toml` is a general build system for all Python projects since PEP-518.  FastAPI uses [flit](https://github.com/takluyver/flit) for packaging.

> minimum pyproject.toml's example

> ```pyproject
> [tool.poetry]
> name = "app"
> version = "0.1.0"
> description = ""
> authors = ["admin <admin@example.com>"]
> 
> [tool.poetry.dependencies]
> python = "^3.9"
> fastapi = "^0.62.0"
> uvicorn = "^0.12.3"
> 
> [tool.poetry.dev-dependencies]
> pytest = "^5.2"
> 
> [build-system]
> requires = ["poetry-core>=1.0.0"]
> build-backend = "poetry.core.masonry.api"
> ```

Also, this should be a poetry example too.

**Comment by Kludex** [comment]
Seems a good opportunity to present this: https://github.com/ycd/manage-fastapi :smile:
